File: 230302-dojo/java/app/src/main/java/Solution.java
//import java.util.*;
//
//class Solution {
//    public int solution(int[] elements) {
//    Set set = new HashSet();
//
//        int length = elements.length;
//
//        // 원소의 갯수
//        for (int i = 1; i <= length; i += 1) {
//            int sum = 0;
//
//            for (int j = 0; j < length; j += 1) {
//                if (j == 0) {
//                    for (int k = 0; k < i; k += 1) {
//                        sum += elements[k];
//                    }
//                }
//
//                if (j != 0) {
//                    sum -= elements[j - 1];
//                    sum += elements[(j + i - 1) % length];
//                }
//                set.add(sum);
//            }
//        }
//        return set.size();
//    }
//}

import java.util.*;

class Solution {
    public int solution(int[] elements) {
        Set set = new HashSet();

        int length = elements.length;
        int[] extended = new int[length * 2];

        for (int i = 0; i < length * 2; i += 1) {
            if (i < length) {
                extended[i] = elements[i];
                continue;
            }

            extended[i] = elements[i - length];
        }

        // 원소 갯수
        for (int i = 1; i <= length; i += 1) {
            int sum = 0;

            for (int j = 0; j < length; j += 1) {
                sum += extended[i + j];
                set.add(sum);
            }
        }

        return set.size();
    }
}

File: 230302-dojo/javascript/solution.js
// function calculateSum(array) {
//   return array.reduce((acc, cur) => acc + cur, 0);
// }

// export default function solution(elements) {
//   const set = new Set();

//   // 원소 갯수를 i
//   for (let i = 1; i <= elements.length; i += 1) {
//     for (let j = 0; j < elements.length; j += 1) {
//       if (j + i <= elements.length) {
//         set.add(calculateSum(elements.slice(j, j + i)));
//         continue;
//       }

//       set.add(calculateSum(elements.slice(j))
//       + calculateSum(elements.slice(0, j + i - elements.length)));
//     }
//   }

//   return set.size;
// }

export default function solution(elements) {
  const set = new Set();

  const subArrayLength = elements.length;

  for (let i = 1; i <= subArrayLength; i += 1) {
    // 슬라이딩 윈도우
    let sum = 0;

    for (let j = 0; j < subArrayLength; j += 1) {
      if (j === 0) {
        sum = elements.slice(0, i).reduce((acc, cur) => acc + cur, 0);
      }

      if (j !== 0) {
        sum -= elements[j - 1];
        sum += elements[(j + i - 1) % subArrayLength];
      }

      set.add(sum);
    }
  }

  return set.size;
}
